gateway/websocket_handlers.go

```diff
@@ -0,0 +1,106 @@
+package gateway
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/buildwithgrove/path/qos/jsonrpc"
+	"github.com/pokt-network/poktroll/pkg/polylog"
+
+	"github.com/buildwithgrove/path/protocol"
+	"github.com/buildwithgrove/path/websockets"
+)
+
+// ---------- Shannon Client Message Handler ----------
+
+var _ websockets.WebSocketMessageHandler = &websocketClientMessageHandler{}
+
+// websocketClientMessageHandler handles websocket client messages
+type websocketClientMessageHandler struct {
+	logger      polylog.Logger
+	protocolCtx ProtocolRequestContext
+	serviceID   protocol.ServiceID
+}
+
+// HandleMessage processes a message from the client.
+func (h *websocketClientMessageHandler) HandleMessage(msgData []byte) ([]byte, error) {
+	logger := h.logger.With("method", "HandleMessage")
+
+	logger.Debug().Msgf("received message from client: %s", string(msgData))
+
+	clientMessageBz, err := h.protocolCtx.SignClientWebsocketMessage(msgData)
+	if err != nil {
+		logger.Error().Err(err).Msg("❌ failed to sign request")
+		return h.getClientErrorResponse(msgData, err)
+	}
+
+	return clientMessageBz, nil
+}
+
+// TODO_IN_THIS_PR(@commoddit): Handle errors using QoS package - ie move JSON-RPC logic to the QoS package.
+
+// ---------- Shannon Error Handling ----------
+
+// getClientErrorResponse sends a JSON-RPC error response to the client
+// Attempts to extract the request ID from the original message for proper error formatting.
+func (h *websocketClientMessageHandler) getClientErrorResponse(
+	originalMessage []byte,
+	err error,
+) ([]byte, error) {
+	logger := h.logger.With("method", "getClientErrorResponse")
+
+	// Try to extract the request ID from the original message
+	var requestID jsonrpc.ID
```

Comment:
Agreed with TODO_IN_THIS_PR above: this logic belongs in the QoS.

Reply:
Per our discussion on Discord I've updated this to a `TODO_TECHDEBT` and we will handle in a subsequent PR.